source/mock/ur_mock_helpers.hpp

```diff
@@ -0,0 +1,108 @@
+/*
+ *
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+ * See LICENSE.TXT
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ * @file ur_mock_helpers.hpp
+ *
+ */
+
+#pragma once
+
+#include "ur_api.h"
+#include <atomic>
+#include <cstddef>
+#include <string>
+#include <unordered_map>
+
+namespace mock {
+
+struct dummy_handle_t_ {
+    dummy_handle_t_() {}
+    std::atomic<size_t> refCounter = 1;
+};
+
+using dummy_handle_t = dummy_handle_t_ *;
+
+// Allocates a dummy handle of type T with support for reference counting.
+// The handle has to be deallocated using 'releaseDummyHandle'.
+template <class T> inline T createDummyHandle() {
+    dummy_handle_t DummyHandlePtr = new dummy_handle_t_();
+    return reinterpret_cast<T>(DummyHandlePtr);
+}
+
+// Decrement reference counter for the handle and deallocates it if the
+// reference counter becomes zero
+template <class T> inline static void releaseDummyHandle(T Handle) {
+    auto DummyHandlePtr = reinterpret_cast<dummy_handle_t>(Handle);
+    const size_t NewValue = --DummyHandlePtr->refCounter;
+    if (NewValue == 0) {
+        delete DummyHandlePtr;
+    }
+}
+
+// Increment reference counter for the handle
+template <class T> inline static void retainDummyHandle(T Handle) {
+    auto DummyHandlePtr = reinterpret_cast<dummy_handle_t>(Handle);
+    ++DummyHandlePtr->refCounter;
+}
+
+struct callbacks {
```

Comment:
I think I'd prefer for this to be a singleton instead of every individual variable being static.